ssl/s23_clnt.c

```diff
@@ -375,12 +357,13 @@ static int ssl23_client_hello(SSL *s)
 
     buf = (unsigned char *)s->init_buf->data;
     if (s->state == SSL23_ST_CW_CLNT_HELLO_A) {
-        /*
-         * Since we're sending s23 client hello, we're not reusing a session, as
-         * we'd be using the method from the saved session instead
-         */
-        if (!ssl_get_new_session(s, 0)) {
-            return -1;
+      /* Check if the session is resumable. If not, drop it. */
+        if (s->session == NULL ||
+            s->session->ssl_version > version ||
+            s->session->session_id_length == 0 ||
+            s->session->not_resumable) {
+            if (!ssl_get_new_session(s, 0))
```

Comment:
Oof. I forgot about that! That was another subtle bit of complexity. If you're not careful, `session->ssl_version` ends up wrong.

At the point that BoringSSL forked, the `s23_clnt.c` code in 1.0.2 defers `ssl_get_new_session` until ServerHello (it actually had it there but `#if 0`'d out because I guess they noticed the problem) while the `s3_clnt.c` code does it early. This meant that, at the time, the DTLS 1.2 logic didn't work because `session->ssl_session` gets filled in before the final version is known.

I decided to defer the `ssl_get_new_session` call until ServerHello, after the version is known. I think this is the most sensible approach. You need a `ssl_get_new_session` call in ServerHello anyway, and this way `session->ssl_version` is correct for free.
https://boringssl.googlesource.com/boringssl/+/81ea0bf538909a0aa3fe2f82726ebdda7dc94ab1

However, it looks like OpenSSL went the other direction (what you did).
https://git.openssl.org/gitweb/?p=openssl.git;a=commit;h=f0348c842e75a77f643e7d34cb88609145b2d320
And them they fix up `ssl_version` later on:
https://git.openssl.org/gitweb/?p=openssl.git;a=commit;h=af3aa2b5ef741a35394c92872cbdbab4d46b9c90
https://git.openssl.org/gitweb/?p=openssl.git;a=commit;h=dc0c888811cebfa2d21c844be0d81335fb2361da

I think that might cause problems. Right now it's fine because this bug means those lines are only reachable when `s->session` is a fresh session. But once we fix this bug, `s->session` might be an established session and then mutating it isn't legal. (You'd be modifying an established session and even introducing race conditions in a multi-threaded program. That's been the subject of a few recent CVEs when the mutations involved pointers.)

Sorry, I'd forgotten how deep this rabbit-hole was! It was all a long time ago for me.